File: src/main/java/com/heybys/oddments/base/jpa/hibernate/LongTypeIdentifierJavaType.java
package com.heybys.oddments.base.jpa.hibernate;

import com.heybys.oddments.base.domain.LongTypeIdentifier;
import java.sql.Types;
import org.hibernate.type.descriptor.WrapperOptions;
import org.hibernate.type.descriptor.java.AbstractClassJavaType;
import org.hibernate.type.descriptor.java.ImmutableMutabilityPlan;
import org.hibernate.type.descriptor.java.MutabilityPlan;
import org.hibernate.type.descriptor.jdbc.JdbcType;
import org.hibernate.type.descriptor.jdbc.JdbcTypeIndicators;

public abstract class LongTypeIdentifierJavaType<T extends LongTypeIdentifier> extends AbstractClassJavaType<T> {

    protected LongTypeIdentifierJavaType(Class<T> clazz) {
        super(clazz);
    }

    @Override
    public String toString(T value) {
        return value == null ? null : value.longValue().toString();
    }

    @Override
    public T fromString(CharSequence string) {
        try {
            return getJavaType().getDeclaredConstructor(Long.class).newInstance(Long.valueOf(string.toString()));
        } catch (Exception ex) {
            throw new IllegalStateException(ex);
        }
    }

    @SuppressWarnings("unchecked")
    public <X> X unwrap(T value, Class<X> type, WrapperOptions options) {
        if (value == null) {
            return null;
        }

        if (Long.class.isAssignableFrom(type)) {
            return (X) value.longValue();
        }

        throw this.unknownUnwrap(type);
    }

    @SuppressWarnings("unchecked")
    public <X> T wrap(X value, WrapperOptions options) {
        if (value == null) {
            return null;
        }

        if (getJavaType().isInstance(value)) {
            return (T) value;
        }

        if (value instanceof Long longValue) {
            try {
                return getJavaType().getDeclaredConstructor(Long.class).newInstance(longValue);
            } catch (Exception e) {
                throw new IllegalStateException(e);
            }
        }

        throw unknownWrap(value.getClass());
    }

    @Override
    public MutabilityPlan<T> getMutabilityPlan() {
        return ImmutableMutabilityPlan.instance();
    }

    @Override
    public JdbcType getRecommendedJdbcType(JdbcTypeIndicators indicators) {
        return indicators.getTypeConfiguration().getJdbcTypeRegistry().getDescriptor(Types.BIGINT);
    }
}

File: src/main/kotlin/com/heybys/oddments/base/jpa/hibernate/LongTypeIdentifierJavaType.kt
package com.heybys.oddments.base.jpa.hibernate

import com.heybys.oddments.base.domain.LongTypeIdentifier
import org.hibernate.type.descriptor.WrapperOptions
import org.hibernate.type.descriptor.java.AbstractClassJavaType
import org.hibernate.type.descriptor.java.ImmutableMutabilityPlan
import org.hibernate.type.descriptor.java.MutabilityPlan
import org.hibernate.type.descriptor.jdbc.JdbcType
import org.hibernate.type.descriptor.jdbc.JdbcTypeIndicators
import java.sql.Types

abstract class LongTypeIdentifierJavaType<T : LongTypeIdentifier>(clazz: Class<T>) :
    AbstractClassJavaType<T>(clazz) {

    override fun toString(value: T?): String? {
        return value?.longValue()?.toString()
    }

    override fun fromString(string: CharSequence?): T {
        return try {
            val longValue =
                string?.toString()?.toLong()
                    ?: throw IllegalArgumentException("Cannot parse null string to Long")
            javaType.getDeclaredConstructor(Long::class.java).newInstance(longValue)
        } catch (ex: Exception) {
            throw IllegalStateException(ex)
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <X> unwrap(value: T?, type: Class<X>, options: WrapperOptions?): X? {
        if (value == null) return null

        return if (Long::class.java.isAssignableFrom(type)) {
            value.longValue() as X
        } else {
            throw unknownUnwrap(type)
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <X> wrap(value: X?, options: WrapperOptions?): T? {
        if (value == null) return null

        return when {
            javaType.isInstance(value) -> value as T
            value is Long ->
                try {
                    javaType.getDeclaredConstructor(Long::class.java).newInstance(value)
                } catch (e: Exception) {
                    throw IllegalStateException(e)
                }
            else -> throw unknownWrap(value.javaClass)
        }
    }

    override fun getMutabilityPlan(): MutabilityPlan<T> {
        return ImmutableMutabilityPlan.instance()
    }

    override fun getRecommendedJdbcType(indicators: JdbcTypeIndicators): JdbcType {
        return indicators.typeConfiguration.jdbcTypeRegistry.getDescriptor(Types.BIGINT)
    }
}
